### PR TITLE
code review

### PR DESCRIPTION
**Pull request by jhcastanod**

### CODE REVIEW

File: src/app/modules/design-patterns/abstract -factory/factories/barcelona-factory.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import { IBuildPlayer } from '../interfaces/build-player.interface';
+import { GoalKeeper } from '../players/goalkeeper';
+import { MidPlayer } from '../players/mid-player';
+import { Striker } from '../players/striker';
+
+export class BarcelonaPlayerBuildFactory implements IBuildPlayer {
+  buildGoalKeeper(): GoalKeeper {
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract -factory/factories/barcelona-factory.ts:7]: you should depends `IGoalKeeper`

File: src/app/modules/design-patterns/abstract -factory/factories/barcelona-factory.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    const player = new GoalKeeper('Ter', 1, '');
+
+    return player;
+  }
+
+  buildMidPlayer(): MidPlayer {
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract -factory/factories/barcelona-factory.ts:13]: you should depends `IMidPlayer`

File: src/app/modules/design-patterns/abstract -factory/factories/barcelona-factory.ts [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    const player = new MidPlayer('Jorge', 7, '');
+
+    return player;
+  }
+
+  buildStrikerPlayer(): Striker {
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract -factory/factories/barcelona-factory.ts:19]: you should depends IStriker

File: src/app/modules/design-patterns/abstract -factory/interfaces/build-player.interface.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import { GoalKeeper } from '../players/goalkeeper';
+import { MidPlayer } from '../players/mid-player';
+import { Striker } from '../players/striker';
+
+export interface IBuildPlayer {
+  buildGoalKeeper(): GoalKeeper;
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract -factory/interfaces/build-player.interface.ts:6]: You should return an interface

File: src/app/modules/design-patterns/abstract -factory/factories/juventus-factory.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import { IBuildPlayer } from '../interfaces/build-player.interface';
+import { GoalKeeper } from '../players/goalkeeper';
+import { MidPlayer } from '../players/mid-player';
+import { Striker } from '../players/striker';
+
+export class JuventusPlayerBuildFactory implements IBuildPlayer {
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract -factory/factories/juventus-factory.ts:6]: The interface should be IBuildPlayerFactory

File: src/app/modules/design-patterns/abstract -factory/factories/juventus-factory.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  buildGoalKeeper(): GoalKeeper {
+    const player = new GoalKeeper('Buffon', 12, '');
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract -factory/factories/juventus-factory.ts:8]: You should create a `JuventusGoalKeeper`

File: src/app/modules/design-patterns/abstract -factory/factories/juventus-factory.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+    return player;
+  }
+
+  buildMidPlayer(): MidPlayer {
+    const player = new MidPlayer('Lucas', 7, '');
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract -factory/factories/juventus-factory.ts:14]: You should create a `JuventusMidPlayer`

File: src/app/modules/design-patterns/abstract -factory/factories/juventus-factory.ts [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+
+    return player;
+  }
+
+  buildStrikerPlayer(): Striker {
+    const player = new Striker('Ronaldo', 7, '...');
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract -factory/factories/juventus-factory.ts:20]: You should create a `JuventusStriker`

File: src/app/modules/design-patterns/abstract -factory/factories/juventus-factory.ts (hunk 6 above)
Comment (luqezman) [reply src/app/modules/design-patterns/abstract -factory/factories/juventus-factory.ts:8]: JuventusGoalKeeper should implements `IGoalKeeper`

File: src/app/modules/design-patterns/abstract -factory/factories/juventus-factory.ts (hunk 7 above)
Comment (luqezman) [reply src/app/modules/design-patterns/abstract -factory/factories/juventus-factory.ts:14]: JuventusGoalKeeper should implements `IMidPlayer`

File: src/app/modules/design-patterns/abstract -factory/factories/juventus-factory.ts (hunk 8 above)
Comment (luqezman) [reply src/app/modules/design-patterns/abstract -factory/factories/juventus-factory.ts:20]: JuventusGoalKeeper should implements `IStrikerPlayer`

File: src/app/app-routing.module.ts [hunk 9]
```diff
@@ -1,7 +1,12 @@
 import { NgModule } from '@angular/core';
 import { Routes, RouterModule } from '@angular/router';
 
-const routes: Routes = [];
+const routes: Routes = [
+  {
+    path: 'team',
```
Comment (luqezman) [inline src/app/app-routing.module.ts:6]: Change to `design-patterns/abstract-factory/team`

File: src/app/modules/design-patterns/abstract-factory/abstract-factory-routing.module.ts [hunk 10]
```diff
@@ -0,0 +1,14 @@
+import { NgModule } from '@angular/core';
+import { Routes, RouterModule } from '@angular/router';
+import { ClubComponent } from './pages/club/club.component';
+
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract-factory/abstract-factory-routing.module.ts:4]: Remove blank ink

File: src/app/modules/design-patterns/abstract-factory/abstract-factory-routing.module.ts [hunk 11]
```diff
@@ -0,0 +1,14 @@
+import { NgModule } from '@angular/core';
+import { Routes, RouterModule } from '@angular/router';
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract-factory/abstract-factory-routing.module.ts:2]: Import keyworkds

Comment (luqezman) [inline src/app/modules/design-patterns/abstract-factory/abstract-factory-routing.module.ts:2]: add blank line here

File: src/app/modules/design-patterns/abstract-factory/abstract-factory-routing.module.ts [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+
+const routes: Routes = [
+  { path: ':club', component: ClubComponent },
+];
+
+@NgModule({
+  imports: [RouterModule.forChild(routes)],
+  exports: [RouterModule]
+})
+export class AbstractFactoryRoutingModule { }
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract-factory/abstract-factory-routing.module.ts:14]: `{ }` to `{}`

File: src/app/modules/design-patterns/abstract-factory/abstract-factory.module.ts [hunk 13]
```diff
@@ -0,0 +1,15 @@
+import { NgModule } from '@angular/core';
+import { CommonModule } from '@angular/common';
+
+import { AbstractFactoryRoutingModule } from './abstract-factory-routing.module';
+import { ClubComponent } from './pages/club/club.component';
+
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract-factory/abstract-factory.module.ts:6]: Remove blank line here

File: src/app/modules/design-patterns/abstract-factory/abstract-factory.module.ts [hunk 14]
```diff
@@ -0,0 +1,15 @@
+import { NgModule } from '@angular/core';
+import { CommonModule } from '@angular/common';
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract-factory/abstract-factory.module.ts:2]: @angular/common should came before @angular/core

File: src/app/modules/design-patterns/abstract-factory/interfaces/goalkeeper-player.interface.ts [hunk 15]
```diff
@@ -0,0 +1,5 @@
+import { AbstractBasePlayer } from './base-player.interface';
+
+export interface IGoalKeeperPlayer extends AbstractBasePlayer {
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract-factory/interfaces/goalkeeper-player.interface.ts:3]: the interface `IGoalKeeperPlayer` should extends IBasePlayer(the base player contract)

Comment (luqezman) [reply src/app/modules/design-patterns/abstract-factory/interfaces/goalkeeper-player.interface.ts:3]: You are doing here Interface segregation

File: src/app/modules/design-patterns/abstract-factory/interfaces/mid-player.interface.ts [hunk 16]
```diff
@@ -0,0 +1,5 @@
+import { AbstractBasePlayer } from './base-player.interface';
+
+export interface IMidPlayer extends AbstractBasePlayer {
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract-factory/interfaces/mid-player.interface.ts:3]: the interface IMidPlayer should extends IBasePlayer(the base player contract)

File: src/app/modules/design-patterns/abstract-factory/interfaces/striker-player.interface.ts [hunk 17]
```diff
@@ -0,0 +1,5 @@
+import { AbstractBasePlayer } from './base-player.interface';
+
+export interface IStrikerPlayer extends AbstractBasePlayer {
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract-factory/interfaces/striker-player.interface.ts:3]: the interface IStrikerPlayer should extends IBasePlayer(the base player contract)

File: src/app/modules/design-patterns/abstract-factory/pages/club/club.component.ts [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+export class ClubComponent implements OnInit {
+  playerBuilderFactory: IBuildPlayerFactory;
+
+  constructor(private readonly activatedRoute: ActivatedRoute) {}
+
+  ngOnInit() {
+    this.activatedRoute.queryParams.subscribe(param => this.choosePlayerBuilderFactory(param));
+  }
+
+  choosePlayerBuilderFactory(team: Params) {
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract-factory/pages/club/club.component.ts:22]: you should create an Factory to decide which factory to use

**Review by luqezman — CHANGES_REQUESTED** [review]

File: src/app/modules/design-patterns/abstract-factory/interfaces/striker-player.interface.ts (hunk 17 above)
Comment (jhcastanod) [reply src/app/modules/design-patterns/abstract-factory/interfaces/striker-player.interface.ts:3]: I don't have IBasePlayer itself... I have `AbstractBasePlayer` which has `name`, `photo`, and `tshirt`.
Do you mean the name of the class or a different class with other properties?

File: src/app/modules/design-patterns/abstract-factory/factories/juventus-factory.ts [hunk 19]
```diff
@@ -0,0 +1,27 @@
+import { IPlayerFactory } from '../interfaces/create-player.interface';
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract-factory/factories/juventus-factory.ts:1]: change file juventus-player-factory.ts

File: src/app/modules/design-patterns/abstract-factory/interfaces/base-player.interface.ts [hunk 20]
```diff
@@ -0,0 +1,17 @@
+export abstract class AbstractBasePlayer {
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract-factory/interfaces/base-player.interface.ts:1]: this file should not belongs to interfaces folder, move to players folders

File: src/app/modules/design-patterns/abstract-factory/interfaces/create-player.interface.ts [hunk 21]
```diff
@@ -0,0 +1,9 @@
+import { IGoalKeeperPlayer } from './goalkeeper-player.interface';
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract-factory/interfaces/create-player.interface.ts:1]: fix filename to player.interface.ts

File: src/app/modules/design-patterns/abstract-factory/factories/barcelona-factory.ts [hunk 22]
```diff
@@ -0,0 +1,27 @@
+import { IPlayerFactory } from '../interfaces/create-player.interface';
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract-factory/factories/barcelona-factory.ts:1]: change filename to barcelona-player-factory.ts

File: src/app/modules/design-patterns/abstract-factory/factories/team-builder-factory.ts [hunk 23]
```diff
@@ -0,0 +1,9 @@
+import { IPlayerFactory } from '../interfaces/create-player.interface';
+import { BarcelonaPlayerFactory } from './barcelona-factory';
+import { JuventusPlayerFactory } from './juventus-factory';
+
+export const teamBuilderFactory = (teamName: 'barcelona' | 'juventus'): IPlayerFactory => {
+  const isBarcelonaTeam = teamName === 'barcelona';
+
+  return isBarcelonaTeam ? new BarcelonaPlayerFactory() : new JuventusPlayerFactory();
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract-factory/factories/team-builder-factory.ts:8]: add this into a variable

File: src/app/modules/design-patterns/abstract-factory/interfaces/base-player.interface.ts (hunk 20 above)
Comment (luqezman) [reply src/app/modules/design-patterns/abstract-factory/interfaces/base-player.interface.ts:1]: fix filename to abstract-base-player.ts

File: src/app/modules/design-patterns/abstract-factory/interfaces/goalkeeper-player.interface.ts [hunk 24]
```diff
@@ -0,0 +1,5 @@
+import { AbstractBasePlayer } from './base-player.interface';
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract-factory/interfaces/goalkeeper-player.interface.ts:1]: change this filename to goalkeeper.interface.ts

File: src/app/modules/design-patterns/abstract-factory/interfaces/mid-player.interface.ts (hunk 16 above)
Comment (luqezman) [inline src/app/modules/design-patterns/abstract-factory/interfaces/mid-player.interface.ts:3]: change from IMidPlayer to IMidFieldPlayer

File: src/app/modules/design-patterns/abstract-factory/interfaces/striker-player.interface.ts [hunk 25]
```diff
@@ -0,0 +1,5 @@
+import { AbstractBasePlayer } from './base-player.interface';
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract-factory/interfaces/striker-player.interface.ts:1]: change filename to striker.interface.ts

**Review by luqezman — CHANGES_REQUESTED** [review]

File: src/app/modules/design-patterns/abstract-factory/interfaces/mid-player.interface.ts (hunk 16 above)
Comment (luqezman) [reply src/app/modules/design-patterns/abstract-factory/interfaces/mid-player.interface.ts:3]: change filename to mid-field.interface.ts

**Review by luqezman — APPROVED** [review]

File: src/app/modules/design-patterns/abstract-factory/pages/club/club.component.ts [hunk 26]
```diff
@@ -0,0 +1,26 @@
...
+})
+export class ClubComponent implements OnInit {
+  players;
+
+  constructor(private readonly activatedRoute: ActivatedRoute) {}
+
+  ngOnInit() {
+    this.activatedRoute.paramMap.subscribe((param: Params) => {
+      const playerFactory = teamBuilderFactory(param.get('club'));
+
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract-factory/pages/club/club.component.ts:18]: remove this line

File: src/app/modules/design-patterns/abstract-factory/pages/club/club.component.ts [hunk 27]
```diff
@@ -0,0 +1,26 @@
...
+import { ActivatedRoute, Params } from '@angular/router';
+import { teamBuilderFactory } from '../../factories/team.factory';
+
+@Component({
+  selector: 'app-club',
+  templateUrl: './club.component.html',
+  styleUrls: ['./club.component.scss']
+})
+export class ClubComponent implements OnInit {
+  players;
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract-factory/pages/club/club.component.ts:11]: add type or add todo

File: src/app/modules/design-patterns/abstract-factory/pages/club/club.component.ts [hunk 28]
```diff
@@ -0,0 +1,26 @@
+import { Component, OnInit } from '@angular/core';
+import { ActivatedRoute, Params } from '@angular/router';
```
Comment (luqezman) [inline src/app/modules/design-patterns/abstract-factory/pages/club/club.component.ts:2]: add blank line here

**Review by luqezman — CHANGES_REQUESTED** [review]

**Pull request merged**